### PR TITLE
[WIP] Add support for access keys

### PR DESCRIPTION
**Pull request by thomas4019**
The main thing missing is some api endpoint for creating access keys.

### CODE REVIEW

File: modules/access_keys/access_keys.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+exports.collections = function () {
+  const access_keys = {
+    _id: 'access_keys',
+    schema: {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        user_id: {
```
Comment (kane-mason) [inline modules/access_keys/access_keys.js:8]: with the more recent changes you will need to add property `user_collection` to cater for users that are not necessarilly `users`

File: auth/index.js [hunk 2]
```diff
@@ -17,6 +17,32 @@ exports.isValidPassword = function (password, hashedPassword) {
...
+  if (key) {
+    if (typeof key == 'string') {
+      const accessKeys = await req.db['access_keys'].find({ key: { $eq: key } })
+      if (accessKeys && accessKeys.length > 0) {
+        const accessKey = accessKeys[0]
+        // check if access key is still valid
+        if (accessKey.expires_at >= new Date().toISOString()) {
+          let user
+          try {
+            user = await req.db['users'].get(accessKey.user_id)
```
Comment (kane-mason) [inline auth/index.js:30]: once `user_collection` added to schema this will become `req.db[accessKey.user_collection]`

File: auth/index.js [hunk 3]
```diff
@@ -17,6 +17,32 @@ exports.isValidPassword = function (password, hashedPassword) {
 exports.doLogin = handler.doLogin
 
 exports.middleware = async function authMiddleware(req, res, next) {
+  const key = req.query['access_key'] || req.headers['x-access-key']
+  if (key) {
+    if (typeof key == 'string') {
+      const accessKeys = await req.db['access_keys'].find({ key: { $eq: key } })
```
Comment (kane-mason) [inline auth/index.js:23]: why dont we short circuit this to only return top 1 result

**Comment by kane-mason** [comment]
> The main thing missing is some api endpoint for creating access keys.

well the standard listeners will exist right? Or do you mean creating an expressa internal listener to populate access key properties?

**Comment by kane-mason** [comment]
> The main thing missing is some api endpoint for creating access keys.

Will these not be added using default collection endpoints? Then we can create in admin?

**Comment by kane-mason** [comment]
@thomas4019 would love to pick this up again and get it in

**Comment by thomas4019** [comment]
@kane-mason Sorry about the delay here. I've merged this in after addressing your comments.

Yeah, you're right that the standard admin endpoints are going to be the way to go for adding, editing and deleting access keys.

**Pull request merged**